src/ucp/core/ucp_context.h

```diff
@@ -13,7 +13,73 @@
+    UCP_MT_TYPE_MUTEX
+} ucp_mt_type_t;
+
+typedef struct ucp_mt_lock {
+    ucp_mt_type_t                 mt_type;
+    union {
+        /* Lock for multithreading support. Either spinlock or mutex is used at
+           at one time. Spinlock is the default option. */
+        pthread_mutex_t           mt_mutex;
+        ucs_spinlock_t            mt_spinlock;
+    } lock;
+} ucp_mt_lock_t;
+
+#if ENABLE_MT
+#define UCP_THREAD_LOCK_INIT(_lock_ptr)                                 \
+    {                                                                   \
+        if ((_lock_ptr)->mt_type == UCP_MT_TYPE_MUTEX) {                \
+            pthread_mutex_init(&((_lock_ptr)->lock.mt_mutex), NULL);    \
+        } else {                                                        \
+            ucs_spinlock_init(&((_lock_ptr)->lock.mt_spinlock));        \
+        }                                                               \
+    }
+#define UCP_THREAD_LOCK_FINALIZE(_lock_ptr)                             \
+    {                                                                   \
+        if ((_lock_ptr)->mt_type == UCP_MT_TYPE_MUTEX) {                \
+            pthread_mutex_destroy(&((_lock_ptr)->lock.mt_mutex));       \
+        } else {                                                        \
+            ucs_spinlock_destroy(&((_lock_ptr)->lock.mt_spinlock));     \
+        }                                                               \
+    }
+#define UCP_THREAD_CS_ENTER(_lock_ptr)                                  \
+    {                                                                   \
+        if ((_lock_ptr)->mt_type == UCP_MT_TYPE_MUTEX) {                \
+            pthread_mutex_lock(&((_lock_ptr)->lock.mt_mutex));          \
+        } else {                                                        \
+            ucs_spin_lock(&((_lock_ptr)->lock.mt_spinlock));            \
+        }                                                               \
+    }
+#define UCP_THREAD_CS_EXIT(_lock_ptr)                                   \
+    {                                                                   \
+        if ((_lock_ptr)->mt_type == UCP_MT_TYPE_MUTEX) {                \
+            pthread_mutex_unlock(&((_lock_ptr)->lock.mt_mutex));        \
+        } else {                                                        \
+            ucs_spin_unlock(&((_lock_ptr)->lock.mt_spinlock));          \
+        }                                                               \
+    }
+#define UCP_THREAD_CS_YIELD(_lock_ptr)                                  \
+    {                                                                   \
+        UCP_THREAD_CS_EXIT(_lock_ptr);                                  \
+        sched_yield();                                                  \
+        UCP_THREAD_CS_ENTER(_lock_ptr);                                 \
+    }
+#define UCP_THREAD_IS_REQUIRED(_lock_ptr) ((_lock_ptr)->mt_type)
```

Comment:
don't need another marco to check if thread required (this results in code bloat)
just need an unconditional versions of THREAD_CS_ENTER/THREAD_CS_EXIT